### PR TITLE
TYP: Typing for ExtensionArray.__getitem__ 

### PR DESCRIPTION
**Pull request by Dr-Irv**
Changes for typing of `ExtensionArray.__getitem__()`.

Required changes to `DateTimeLikeArrayMixin` for correct compatibility.

### CODE REVIEW

File: pandas/core/arrays/_mixins.py [hunk 1]
```diff
@@ -276,9 +277,19 @@ def __setitem__(self, key, value):
...
+    @overload
+    def __getitem__(
+        self: NDArrayBackedExtensionArrayT, item: slice | np.ndarray | Sequence[int]
+    ) -> NDArrayBackedExtensionArrayT:
+        ...
+
     def __getitem__(
         self: NDArrayBackedExtensionArrayT,
-        key: PositionalIndexer2D,
+        key: PositionalIndexer,
```
Comment (jbrockmendel) [inline pandas/core/arrays/_mixins.py:292]: why is this changed?  (same question in datetimelike?)  PositionalIndexer2D is correct

Comment (Dr-Irv) [reply pandas/core/arrays/_mixins.py:292]: I'm not so sure that it is correct.  The difference between `PostionalIndexer` and `PositionalIndexer2D` is that the latter includes accepting `Tuple[PositionalIndexer, PositionalIndexer]` as an argument.  For an `ExtensionArray`, those are always one-dimensional, and don't accept 2 dimensional indexers.  

For `datetimelike` and the mixins, I don't think they take 2D indexers as well.

If they do, then we are violating the Liskov substitution principle, in that the subclass implementation of `__getitem__()` is accepting arguments beyond what the (one dimensional) `ExtensionArray` can accept.

Comment (jbrockmendel) [reply pandas/core/arrays/_mixins.py:292]: > For datetimelike and the mixins, I don't think they take 2D indexers as well.

NDArrayBackedExtensionArray can be 2D, so can take 2D indexers.  ATM Categorical is the only subclass that still prohibits 2D.

Comment (Dr-Irv) [reply pandas/core/arrays/_mixins.py]: OK, @jbrockmendel I've changed the base class to use `PositionalIndexer2D` .  But mypy is complaining about using `super().__getitem__(key)` in `datetimelike.py` in `__getitem__()`, and I've put in a type ignore for that, because I think mypy is incorrect.  I couldn't find any other workaround given the error message from mypy

File: pandas/core/arrays/_mixins.py [hunk 2]
```diff
@@ -276,10 +277,20 @@ def __setitem__(self, key, value):
...
 
+    @overload
+    def __getitem__(self, item: int | np.integer) -> Any:
+        ...
+
+    @overload
     def __getitem__(
-        self: NDArrayBackedExtensionArrayT,
+        self: NDArrayBackedExtensionArray, item: slice | np.ndarray | Sequence[int]
+    ) -> NDArrayBackedExtensionArray:
```
Comment (jbrockmendel) [inline pandas/core/arrays/_mixins.py:287]: is there a TypeVar for this?

File: pandas/core/arrays/datetimelike.py [hunk 3]
```diff
@@ -309,14 +309,28 @@ def __array__(self, dtype: NpDtype | None = None) -> np.ndarray:
...
         """
         This getitem defers to the underlying array, which by-definition can
         only handle list-likes, slices, and integer scalars
         """
-        result = super().__getitem__(key)
+        #  error: Invalid index type "Union[Union[int, integer[Any], slice,
+        # Sequence[int], ndarray], Tuple[Union[int, integer[Any], slice, Sequence[int],
+        # ndarray], Union[int, integer[Any], slice, Sequence[int], ndarray]]]" for
+        # "DatetimeLikeArrayMixin"; expected type "Union[int, integer[Any]]"
+        result = super().__getitem__(key)  # type: ignore[index]
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimelike.py:333]: best guess is we need _libs/arrays.pyi

File: pandas/core/arrays/_mixins.py (hunk 2 above)
Comment (Dr-Irv) [reply pandas/core/arrays/_mixins.py:287]: > is there a TypeVar for this?

Yes, there is, but there is no need to use it.

File: pandas/core/arrays/datetimelike.py (hunk 3 above)
Comment (Dr-Irv) [reply pandas/core/arrays/datetimelike.py:333]: > best guess is we need _libs/arrays.pyi

But `_libs/arrays.pyx` doesn't define `__getitem__`.  What I can't figure out is why it thinks it is expecting `Union[int, integer[Any]]` when the `super()` should be referring to `NDArrayBackedExtensionArray` , which is a subclass of `ExtensionArray`.  Even if I use `super(NDArrayBackedExtensionArray, self)` I get the same mypy message.

**Comment by Dr-Irv** [comment]
@simonjayhawkins can you review, please?

File: pandas/core/arrays/_mixins.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/_mixins.py]: > If they do, then we are violating the Liskov substitution principle, in that the subclass implementation of `__getitem__()` is accepting arguments beyond what the (one dimensional) `ExtensionArray` can accept.

I think that's OK, subclass can accept wider types and additional parameters and return narrower types.

Comment (simonjayhawkins) [reply pandas/core/arrays/_mixins.py]: this passes mypy checks

```
from __future__ import annotations


class Foo:
    def bar(self, param: int) -> int | str:
        pass


class FooSubclass(Foo):
    def bar(self, param: int | str, extra=int) -> str:
        pass
```

and this fails

```
from __future__ import annotations


class Foo:
    def bar(self, param: int | str, param2=int) -> int:
        pass


class FooSubclass(Foo):
    def bar(self, param: int) -> int | str:
        pass
```

File: pandas/core/arrays/base.py [hunk 4]
```diff
@@ -26,7 +27,7 @@
     ArrayLike,
     Dtype,
     FillnaOptions,
-    PositionalIndexer,
+    PositionalIndexer2D,
```
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:30]: revert this

File: pandas/core/arrays/_mixins.py [hunk 5]
```diff
@@ -204,10 +205,20 @@ def __setitem__(self, key, value):
     def _validate_setitem_value(self, value):
         return value
 
+    @overload
+    def __getitem__(self, item: int | np.integer) -> Any:
```
Comment (simonjayhawkins) [inline pandas/core/arrays/_mixins.py:209]: item or key? item would be consistent with base class, key would be consistent with non-overload signature.

File: pandas/core/arrays/base.py [hunk 6]
```diff
@@ -291,8 +292,15 @@ def _from_factorized(cls, values, original):
...
+    @overload
+    def __getitem__(self, item: int | np.integer) -> Any:
+        ...
 
-    def __getitem__(self, item: PositionalIndexer) -> ExtensionArray | Any:
+    @overload
+    def __getitem__(self, item: slice | np.ndarray | Sequence[int]) -> ExtensionArray:
+        ...
+
+    def __getitem__(self, item: PositionalIndexer2D) -> ExtensionArray | Any:
```
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:303]:
```suggestion
    def __getitem__(
        self: ExtensionArrayT, item: slice | np.ndarray | Sequence[int]
    ) -> ExtensionArrayT:
        ...

    def __getitem__(
        self: ExtensionArrayT, item: PositionalIndexer
    ) -> ExtensionArrayT | Any:
```

and the ignore in pandas/core/arrays/_mixins.py can be removed

Comment (simonjayhawkins) [reply pandas/core/arrays/base.py:303]: Also the docstring does not mention sequence. should maybe update for consistency although we shouldn't change the EA base class api, so maybe we should not include Sequence in the allowed types here?

File: pandas/core/arrays/_mixins.py [hunk 7]
```diff
@@ -204,10 +205,20 @@ def __setitem__(self, key, value):
...
-        self: NDArrayBackedExtensionArrayT,
+        self: NDArrayBackedExtensionArray, item: slice | np.ndarray | Sequence[int]
+    ) -> NDArrayBackedExtensionArray:
+        ...
+
+    def __getitem__(
+        self: NDArrayBackedExtensionArray,
         key: PositionalIndexer2D,
-    ) -> NDArrayBackedExtensionArrayT | Any:
+    ) -> NDArrayBackedExtensionArray | Any:
```
Comment (simonjayhawkins) [inline pandas/core/arrays/_mixins.py:221]: revert this

File: pandas/core/arrays/_mixins.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/_mixins.py:215]: please do this for consistency

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @Dr-Irv generally lgtm. issue with api regarding using Sequence that may require some discussion

File: pandas/core/arrays/datetimelike.py [hunk 8]
```diff
@@ -308,14 +308,28 @@ def __array__(self, dtype: NpDtype | None = None) -> np.ndarray:
...
+    @overload
+    def __getitem__(
+        self: DatetimeLikeArrayT, item: slice | np.ndarray | Sequence[int]
+    ) -> DatetimeLikeArrayT:
+        ...
+
     def __getitem__(
-        self, key: PositionalIndexer2D
-    ) -> DatetimeLikeArrayMixin | DTScalarOrNaT:
+        self: DatetimeLikeArrayT, key: PositionalIndexer2D
```
Comment (simonjayhawkins) [inline pandas/core/arrays/datetimelike.py:322]:
```suggestion
        self: DatetimeLikeArrayT, key: PositionalIndexer
```

and remove the ignore.

**Comment by simonjayhawkins** [comment]
> Thanks @Dr-Irv generally lgtm. issue with api regarding using Sequence that may require some discussion

Thinking some more, probably best to remove Sequence altogether here.

A tuple of ints is a Sequence[int] and returns a Scalar on our 1d arrays in some cases and errors in others, so we need to add an extra overload for tuple to make the the annotations correct. This will then likely create LSP issues with the subclasses that can be backed by a 2d array as the return type of and tuple of ints will be inconsistent will the base implementation.

File: pandas/core/arrays/datetimelike.py (hunk 8 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/datetimelike.py:322]: @Dr-Irv the non-overloaded signature is used to check the function itself.

mypy will use the overloads to check consistency elsewhere. To allow PositionalIndexer2D to be used, need to add the Tuple types to the overloads.

**Review by simonjayhawkins — COMMENTED** [review]
@Dr-Irv can you also add annotations to the other array subclasses, Categorical, Interval, Masked, ArrowStringArray and Sparse before catering for the 2d array subclasses

File: pandas/core/arrays/base.py (hunk 6 above)
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:303]: I think it is better to change the API.  We always accept a `Sequence[int]` 

> > Thanks @Dr-Irv generally lgtm. issue with api regarding using Sequence that may require some discussion
> 
> Thinking some more, probably best to remove Sequence altogether here.
> 
> A tuple of ints is a Sequence[int] and returns a Scalar on our 1d arrays in some cases and errors in others, so we need to add an extra overload for tuple to make the the annotations correct. This will then likely create LSP issues with the subclasses that can be backed by a 2d array as the return type of and tuple of ints will be inconsistent will the base implementation.

I've changed it to `list[int]` , because we do accept that, and should reflect that in the EA API

File: pandas/core/arrays/base.py (hunk 4 above)
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:30]: done in next commit

File: pandas/core/arrays/_mixins.py (hunk 7 above)
Comment (Dr-Irv) [reply pandas/core/arrays/_mixins.py:221]: done in next commit

File: pandas/core/arrays/_mixins.py (hunk 2 above)
Comment (Dr-Irv) [reply pandas/core/arrays/_mixins.py:215]: done in next commit

File: pandas/core/arrays/datetimelike.py (hunk 3 above)
Comment (Dr-Irv) [reply pandas/core/arrays/datetimelike.py:332]: Fixed with a `cast`

File: pandas/core/arrays/_mixins.py (hunk 5 above)
Comment (Dr-Irv) [reply pandas/core/arrays/_mixins.py:209]: fixed in next commit

File: pandas/core/arrays/datetimelike.py (hunk 8 above)
Comment (Dr-Irv) [reply pandas/core/arrays/datetimelike.py:322]: fixed in next commit

**Comment by Dr-Irv** [comment]
> @Dr-Irv can you also add annotations to the other array subclasses, Categorical, Interval, Masked, ArrowStringArray and Sparse before catering for the 2d array subclasses

Done in next commit.  Created issues in the block and array managers where now the `type ignore` comments have to change, as we are strictly defining EA's to be 1D.  So I changed those comments.

**Comment by Dr-Irv** [comment]
@simonjayhawkins  pushed changes a week ago, so looking forward to a new review.

File: pandas/_typing.py [hunk 9]
```diff
@@ -211,7 +211,6 @@
 # TODO: add Ellipsis, see
 # https://github.com/python/typing/issues/684#issuecomment-548203158
 # https://bugs.python.org/issue41810
-PositionalIndexer = Union[int, np.integer, slice, Sequence[int], np.ndarray]
-PositionalIndexer2D = Union[
-    PositionalIndexer, Tuple[PositionalIndexer, PositionalIndexer]
-]
+PositionalIndexer = Union[int, np.integer, slice, List[int], np.ndarray]
```
Comment (jbrockmendel) [inline pandas/_typing.py:214]: can you add a comment as to why this is List[int] instead of Sequence[int]?  I almost asked about it before reading back through the thread

File: pandas/core/arrays/_mixins.py [hunk 10]
```diff
@@ -204,6 +206,17 @@ def __setitem__(self, key, value):
     def _validate_setitem_value(self, value):
         return value
 
+    @overload
+    def __getitem__(self, key: int | np.integer) -> Any:
```
Comment (jbrockmendel) [inline pandas/core/arrays/_mixins.py:210]: not for this PR, but could/should we do a Generic think to tighten this Any?

File: pandas/core/arrays/_mixins.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/_mixins.py:217]: did the commit using the TypeVar not get pushed?

File: pandas/core/arrays/categorical.py [hunk 11]
```diff
@@ -2015,7 +2018,20 @@ def __repr__(self) -> str:
...
+    @overload
+    def __getitem__(
+        self: CategoricalT,
+        key: slice | np.ndarray | list[int] | PositionalIndexerTuple,
+    ) -> CategoricalT:
+        ...
+
+    def __getitem__(
+        self: CategoricalT, key: PositionalIndexer2D
+    ) -> CategoricalT | object:
```
Comment (jbrockmendel) [inline pandas/core/arrays/categorical.py:2034]: object-> Any?

File: pandas/core/arrays/datetimelike.py [hunk 12]
```diff
@@ -308,14 +309,26 @@ def __array__(self, dtype: NpDtype | None = None) -> np.ndarray:
...
+    ) -> DatetimeLikeArrayT | DTScalarOrNaT:
         """
         This getitem defers to the underlying array, which by-definition can
         only handle list-likes, slices, and integer scalars
         """
-        result = super().__getitem__(key)
+        # Use cast as we know we will get back a DatetimeLikeArray
+        result = cast(DatetimeLikeArrayT, super().__getitem__(key))
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimelike.py:331]: i dont understand.  in the next line we're checking if the result is a scalar, so why are you sure this is non-scalar here?

File: pandas/core/arrays/sparse/array.py [hunk 13]
```diff
@@ -77,6 +81,17 @@
 
 import pandas.io.formats.printing as printing
 
+# See https://github.com/python/typing/issues/684
```
Comment (jbrockmendel) [inline pandas/core/arrays/sparse/array.py:84]: why doing this here and not in pd._typing?

File: pandas/core/arrays/string_arrow.py [hunk 14]
```diff
@@ -341,7 +342,19 @@ def _concat_same_type(cls, to_concat) -> ArrowStringArray:
             )
         )
 
-    def __getitem__(self, item: PositionalIndexer) -> Any:
+    @overload
+    def __getitem__(self, item: int | np.integer) -> str:
```
Comment (jbrockmendel) [inline pandas/core/arrays/string_arrow.py:346]: i dont know this class that well; could this give pd.NA?

File: pandas/_typing.py (hunk 9 above)
Comment (Dr-Irv) [reply pandas/_typing.py:214]: done in next commit

File: pandas/core/arrays/_mixins.py (hunk 10 above)
Comment (Dr-Irv) [reply pandas/core/arrays/_mixins.py:210]: Possibly.  Should I create an issue for discussion?

File: pandas/core/arrays/categorical.py (hunk 11 above)
Comment (Dr-Irv) [reply pandas/core/arrays/categorical.py:2034]: Can a categorical be based on a List?

File: pandas/core/arrays/datetimelike.py (hunk 12 above)
Comment (Dr-Irv) [reply pandas/core/arrays/datetimelike.py:331]: I fixed this in the next commit.  Now two casts are needed in the implementation.

File: pandas/core/arrays/string_arrow.py (hunk 14 above)
Comment (Dr-Irv) [reply pandas/core/arrays/string_arrow.py:346]: Yes, fixed in next commit.

File: pandas/core/arrays/sparse/array.py (hunk 13 above)
Comment (Dr-Irv) [reply pandas/core/arrays/sparse/array.py:84]: This is the only place where we use `ellipsis` as a possible parameter to a method, which is why this construction is needed.  I can move it if you want.

Comment (jbrockmendel) [reply pandas/core/arrays/sparse/array.py:84]: there's a TODO in pd._typing about adding Ellipsis to PositionalIndexer (also references the same GH issue as you did here)

Comment (Dr-Irv) [reply pandas/core/arrays/sparse/array.py:84]: > there's a TODO in pd._typing about adding Ellipsis to PositionalIndexer (also references the same GH issue as you did here)

Fair enough, but I think we should handle `Ellipsis` in a separate PR, as I'm guessing there will be other things we have to fix if we add `Ellipsis` to `PositionalIndexer` in  `pd._typing`

Comment (jbrockmendel) [reply pandas/core/arrays/sparse/array.py:84]: sure

**Comment by Dr-Irv** [comment]
@simonjayhawkins @jbrockmendel Based on our discussion today, I think this is where we want to be.  The base EA class `getitem()` is accepting single dimensional arguments, but the mixin classes accept 2D versions, including tuples.

Can you do a pass through the current version and see if there is anything else to change?

File: pandas/core/arrays/_mixins.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/_mixins.py:217]: NDArrayBackedExtensionArrayT?

File: pandas/core/arrays/categorical.py (hunk 11 above)
Comment (jbrockmendel) [reply pandas/core/arrays/categorical.py:2034]: im not clear what "based on" means here

File: pandas/core/arrays/interval.py [hunk 15]
```diff
@@ -629,7 +631,19 @@ def __iter__(self):
     def __len__(self) -> int:
         return len(self._left)
 
-    def __getitem__(self, key):
+    @overload
+    def __getitem__(self, key: int | np.integer) -> Interval:
```
Comment (jbrockmendel) [inline pandas/core/arrays/interval.py:635]: can also be np.nan i think

File: pandas/core/arrays/_mixins.py (hunk 2 above)
Comment (Dr-Irv) [reply pandas/core/arrays/_mixins.py:217]: fixed in next commit.  Not sure how I missed that.

File: pandas/core/arrays/categorical.py (hunk 11 above)
Comment (Dr-Irv) [reply pandas/core/arrays/categorical.py:2034]: If we use `Any`, then that means any type and mypy stops further type checking.  From some other PR, the idea here is that we can return an `object`, which means other type checking would work downstream.

My comment about "categorical be based on a List" was misleading.

File: pandas/core/arrays/interval.py (hunk 15 above)
Comment (Dr-Irv) [reply pandas/core/arrays/interval.py:635]: Not for `IntervalArray` or any other `ExtensionArray` :
```python
>>> pd.array([pd.Interval(1,2), pd.Interval(3,4)])[np.nan]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "c:\Code\pandas_dev\pandastypes\pandas\core\arrays\interval.py", line 648, in __getitem__
    left = self._left[key]
IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices
```

```python
>>> pd.array([1,2,3])[np.nan]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "c:\Code\pandas_dev\pandastypes\pandas\core\arrays\masked.py", line 159, in __getitem__
    return type(self)(self._data[item], self._mask[item])
IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices
```

Comment (jbrockmendel) [reply pandas/core/arrays/interval.py:634]: i meant it can return np.nan

Comment (Dr-Irv) [reply pandas/core/arrays/interval.py:634]: got it.  Fixed in next commit

File: pandas/core/arrays/categorical.py (hunk 11 above)
Comment (Dr-Irv) [reply pandas/core/arrays/categorical.py:2020]: for consistency with other places where I did `__getitem__` overloads, I changed this to `Any`

**Comment by Dr-Irv** [comment]
@simonjayhawkins ping...

**Comment by Dr-Irv** [comment]
pinging @simonjayhawkins

File: pandas/core/arrays/_mixins.py (hunk 10 above)
Comment (jreback) [reply pandas/core/arrays/_mixins.py:191]: isn't tihs a PositionScalar ?

File: pandas/core/arrays/_mixins.py [hunk 16]
```diff
@@ -185,6 +187,17 @@ def __setitem__(self, key, value):
...
+        ...
+
+    @overload
+    def __getitem__(
+        self: NDArrayBackedExtensionArrayT,
+        key: slice | np.ndarray | list[int] | PositionalIndexerTuple,
```
Comment (jreback) [inline pandas/core/arrays/_mixins.py:197]: isn't this PostionalIndexer2D

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/_mixins.py (hunk 16 above)
Comment (Dr-Irv) [reply pandas/core/arrays/_mixins.py:197]: First, there is no `PositionScalar`, so I'm not sure what you are referring to.

For the `NDArrayBackedExtensionArray`, which extends `ExtensionArray`, `__getitem__()` needs to handle the `__getitem__()` arguments for `ExtensionArray`, which is `PositionalIndexer` , and then `NDArrayBackedExtensionArray` can widen the accepted arguments.   We then have

```python
PositionalIndexer = Union[int, np.integer, slice, List[int], np.ndarray]
PositionalIndexerTuple = Tuple[PositionalIndexer, PositionalIndexer]
PositionalIndexer2D = Union[PositionalIndexer, PositionalIndexerTuple]
```

For `NDArrayBackedExtensionArray.__getitem__()`, it accepts `PositionalIndexer2D`.  The overloads then indicate what are the return types for the various subtypes of `PositionalIndexer2D`.  This has separate return types if the arguments are `int | np.integer` versus `slice | np.ndarray | list[int] | PositionalIndexerTuple` .

Comment (jreback) [reply pandas/core/arrays/_mixins.py:197]: oh i had edited to `PositionalIndexer2D` which looks like the correct return type here

Comment (Dr-Irv) [reply pandas/core/arrays/_mixins.py:197]: > oh i had edited to `PositionalIndexer2D` which looks like the correct return type here

You mean "argument type", not "return type" ??

Here's the full code:
```python
    @overload
    def __getitem__(self, key: int | np.integer) -> Any:
        ...

    @overload
    def __getitem__(
        self: NDArrayBackedExtensionArrayT,
        key: slice | np.ndarray | list[int] | PositionalIndexerTuple,
    ) -> NDArrayBackedExtensionArrayT:
        ...

    def __getitem__(
        self: NDArrayBackedExtensionArrayT,
        key: PositionalIndexer2D,
    ) -> NDArrayBackedExtensionArrayT | Any:
```
The way it works is as follows with respect to the type of `key`:

- if the type is `int | np.integer`, return type is `Any`
- if the type is `slice | np.ndarray | list[int] | PositionalIndexerTuple`, return type is definitely `NDArrayBackedExtensionArrayT`
- If the type is `PositionalIndexer`, return type is either `NDArrayBackedExtensionArrayT | Any`

This makes some downstream typing work correctly by creating the separation this way.

**Comment by Dr-Irv** [comment]
weekly ping @simonjayhawkins

**Comment by Dr-Irv** [comment]
another ping @simonjayhawkins

**Comment by jreback** [comment]
same @Dr-Irv 

cc @twoertwein if you can have a look as you are looking at similar things

File: pandas/core/arrays/categorical.py [hunk 17]
```diff
@@ -2003,7 +2007,18 @@ def __repr__(self) -> str:
...
 
-    def __getitem__(self, key):
+    @overload
+    def __getitem__(self, key: int | np.integer) -> Any:
+        ...
+
+    @overload
+    def __getitem__(
+        self: CategoricalT,
+        key: slice | np.ndarray | list[int] | PositionalIndexerTuple,
```
Comment (twoertwein) [inline pandas/core/arrays/categorical.py:2017]: Would it be helpful to create a union for `slice | np.ndarray | list[int]`? There are a few places where we want to allow non-scalar/vector indices, e.g., for [take](https://github.com/pandas-dev/pandas/pull/43418).

I'm fine with using `list[int]`, some people might prefer `Sequence[int]` (if any Sequence is supported).

File: pandas/core/arrays/datetimelike.py [hunk 18]
```diff
@@ -313,17 +314,33 @@ def __array__(self, dtype: NpDtype | None = None) -> np.ndarray:
             return np.array(list(self), dtype=object)
         return self._ndarray
 
+    @overload
+    def __getitem__(self, item: int | np.integer) -> DTScalarOrNaT:
```
Comment (twoertwein) [inline pandas/core/arrays/datetimelike.py:318]: maybe something like `ScalarIndex`

File: pandas/core/arrays/categorical.py (hunk 17 above)
Comment (Dr-Irv) [reply pandas/core/arrays/categorical.py:2017]: `Sequence[int]` won't work because we don't want to support tuples here.

I'll create the other unions in `__typing__.py`

File: pandas/core/arrays/datetimelike.py (hunk 18 above)
Comment (Dr-Irv) [reply pandas/core/arrays/datetimelike.py:318]: Created a `ScalarIndexer` in next commit

File: pandas/_typing.py [hunk 19]
```diff
@@ -209,7 +209,9 @@
 # TODO: add Ellipsis, see
 # https://github.com/python/typing/issues/684#issuecomment-548203158
 # https://bugs.python.org/issue41810
-PositionalIndexer = Union[int, np.integer, slice, Sequence[int], np.ndarray]
-PositionalIndexer2D = Union[
-    PositionalIndexer, Tuple[PositionalIndexer, PositionalIndexer]
-]
+# Using List[int] here rather than Sequence[int] to disallow tuples.
```
Comment (jreback) [inline pandas/_typing.py:212]: can you followup and give some hints on where these should be used (a.g. a line of description for each; example....)

**Review by jreback — APPROVED** [review]

File: pandas/_typing.py (hunk 19 above)
Comment (Dr-Irv) [reply pandas/_typing.py:216]: > can you followup and give some hints on where these should be used (a.g. a line of description for each; example....)

Done in a new commit - added comments - let me know if you want more

Comment (jreback) [reply pandas/_typing.py:216]: got it, looks good.

**Comment by Dr-Irv** [comment]
@jreback all green now (except for the one test on Windows that always seems to time out)

**Comment by jreback** [comment]
thanks @Dr-Irv
